lib/build_cache.py

```diff
@@ -744,11 +744,37 @@ def configure(self):
       ch.close_(fp)
       # Ignore list entries:
       #
-      # 1. Git has no default gitignore, but cancel any global gitignore rules
-      #    the user might have. https://stackoverflow.com/a/26681066
+      #   1. Git has no default gitignore, but cancel any global gitignore rules
+      #      the user might have. https://stackoverflow.com/a/26681066
       #
-      # 2. The oddly-named GIT_DIR.
+      #   2. The oddly-named GIT_DIR.
+      #
+      # It it easier to just write list we want every time, rather than trying
+      # to figure out if an update is needed.
       (self.root // "info/exclude").file_write("!*\n%s\n" % im.GIT_DIR)
+      # Remove old .gitignore files from all commits. While there are nice
+      # tools to do this (e.g. “git filter-repo”), we don’t want to depend on
+      # an external tool. Thus, the options seem to be “filter-branch” or
+      # “export” followed by “import”. Around half of the “filter-branch” man
+      # page is devoted to explaining why not to use it, so we use the latter.
+      #
+      # NOTE: Without --reflog, “fast-export” will omit commits on unnamed
+      # branches (i.e., accessible only via reflog), but with it, we get
+      # “commit” instructions in the stream with no branch name, which
+      # “fast-import” won’t accept. Therefore, we just delete those commits,
+      # to keep bad .gitignore from creeping back in. (I couldn’t figure out
+      # how to fix this for “filter-branch” either, which I didn’t want to use
+      # anyway for the reasons above.)
+      if (ch.storage.bucache_needs_ignore_upgrade.exists_()):
+         ch.INFO("upgrading build cache to v6+, some cached data may be lost",
+                 "see release notes for v0.32")
+         old = self.git(["fast-export", "--no-data", "--", "--all"]).stdout
+         new = re.sub(r"^(D|M [0-7]+ [0-9a-f]+) \.(git|weirdal_)ignore$",
+                      "#\g<0>", old, flags=re.MULTILINE)
+         #fs.Path("/tmp/new").file_write(new)
```

Comment:
do you want this comment?

Reply:
yeah, it was to show how to debug since it wasn't a standard method